### PR TITLE
feat: [ios] video codec option

### LINKED ISSUE #485: 🐛 iPhone 11 doesn't record video with H.264 codec
### What were you trying to do?

I tried to record video with H.264 codec

### Reproduceable Code

```tsx
<Camera
    ref={cameraRef}
    device={devices.front}
    video={true}
    audio={true}
    isActive={true}
    preset="iframe-960x540"
    fps={25}
    onInitialized={() => setCameraReady(true)}
/>
```

### What happened instead?

When I tried to record a video with H.264 codec, I got HEVC video codec.

### Relevant log output

_No response_

### Device

iPhone 11 (IOS 15.0.1)

### VisionCamera Version

2.8.0.

### Additional information

- [ ] I am using Expo
- [X] I have read the [Troubleshooting Guide](https://mrousavy.github.io/react-native-vision-camera/docs/guides/troubleshooting)
- [X] I agree to follow this project's [Code of Conduct](https://github.com/mrousavy/react-native-vision-camera/blob/main/CODE_OF_CONDUCT.md)
- [X] I searched for [similar issues in this repository](https://github.com/mrousavy/react-native-vision-camera/issues) and found none.

### PR DESCRIPTION
**Pull request by paulrostorp**
## What

For IOS only, adds an optional `videoCodec` parameter to the `startRecording` options to specify the codec to record with. Requires IOS 11 (see commit a09201c2bc88eb6c11417046824cac88d9338069 vs 17b960603298e19d45e10933fe6dbf1da6ee00a6 .... please advise if not suitable)

## Changes
- handle user provided video codec in `CameraView+RecordVideo.swift`
- adds an optional `videoCodec` parameter to the `RecordVideoOptions` interface
- adds a `getAvailableVideoCodecs` method which takes a `fileType` parameter (see: https://stackoverflow.com/questions/46843849/why-is-videodataoutputs-availablevideocodectypes-empty/58017253)

## Tested on
- iphone 7, ios 12
- ipad air, ios 14.4

## Related issues

resolves #485

### CODE REVIEW

File: ios/CameraView+RecordVideo.swift [hunk 1]
```diff
@@ -108,12 +108,20 @@ extension CameraView: AVCaptureVideoDataOutputSampleBufferDelegate, AVCaptureAud
...
+      if let codecString = options["videoCodec"] as? String {
+        videoCodec = AVVideoCodecType(withString: codecString)
+      }
+
       // Init Video
-      guard let videoSettings = videoOutput.recommendedVideoSettingsForAssetWriter(writingTo: fileType),
-            !videoSettings.isEmpty else {
+      guard let videoSettings = videoCodec != nil ?
+        videoOutput.recommendedVideoSettings(forVideoCodecType: videoCodec!, assetWriterOutputFileType: fileType) :
+        videoOutput.recommendedVideoSettingsForAssetWriter(writingTo: fileType),
```
Comment (mrousavy) [inline ios/CameraView+RecordVideo.swift]: This is too much complexity, try to lift this out of that guard to keep it easy to read.

File: src/VideoFile.ts [hunk 2]
```diff
@@ -21,6 +32,13 @@ export interface RecordVideoOptions {
    * Called when the recording has been successfully saved to file.
    */
   onRecordingFinished: (video: VideoFile) => void;
+  /**
+   * IOS only
+   * Specify the video codec to use. To get a list of available video codecs use the `getAvailableVideoCodecs()` function.
```
Comment (mrousavy) [inline src/VideoFile.ts]:
```suggestion
   * Set the video codec to record in. Different video codecs affect video quality and video size.
   * To get a list of all available video codecs use the `getAvailableVideoCodecs()` function.
```

**Review by mrousavy — CHANGES_REQUESTED** [review]
Hey Paul, thank you for your PR! 🖤 

VisionCamera's philosophy is to avoid platform-specifics and provide a uniform abstraction that doesn't require the user to understand the native underlying Camera architectures to use VisionCamera. 
Can you add the Video Codec option to Android as well? The latest CameraX version ships with a separate Video Recording library, maybe that allows Video Codec? Or via Camera2 interop.

Also I left two small comments, otherwise it looks good :)

File: src/VideoFile.ts (hunk 2 above)
Comment (paulrostorp) [reply src/VideoFile.ts]: This description is wrong, that method isn't actually available, I made a copy-paste error when copying from https://github.com/mrousavy/react-native-vision-camera/pull/2/files#diff-432ccbc3244e65723bf8031a79bbad1191b792c59affdbd2ea93ef6e6b1138d4L6

**Comment by paulrostorp** [comment]
@mrousavy Thanks for your review, I've made some changes according to it, incl. adding a `getAvailableVideoCodecs` method.

I, definitely, agree with providing cross-platform features, however, after spending a few hours looking into it, it doesn't seem that the `Camera2interop` or the `Recorder` from the latest cameraX expose functionality to specify the codec or to retrieve available codecs.

The only thing I am able to find is using the `MediaRecorder`(https://developer.android.com/reference/android/media/MediaRecorder) but that doesn't seem suitable...

My native android skills are almost non-existent, so I'm quite unsure about this, but to support these findings, I couldn't find any other React Native (or flutter) packages that support choosing a codec on android.

**Comment by mrousavy** [comment]
Nice job Paul! Thank you for your PR. 🖤  Let's ship it 🚀

**Review by mrousavy — APPROVED** [review]

**Pull request merged**